src/language.rs

```diff
@@ -4,10 +4,22 @@ use {
     regex::Regex,
     std::collections::HashMap,
     strum::{EnumIter, EnumString},
+    std::env,
 };
 
+#[derive(Clone)]
+pub struct Colors {
+    basic_colors: Vec<Color>,
+    true_colors: Option<Vec<Color>>,
+}
+
+macro_rules! define_colors {
+  ( vec![ $( $color:expr ),+ ] ) => { Colors { basic_colors: vec![$( $color ),+ ], true_colors: None } };
+  ( $color:expr ) => { $color };
+}
+
 macro_rules! define_languages {
-    ($( { $name:ident, $ascii:literal, $display:literal, $colors:expr $(, $serialize:literal )? } ),* ,) => {
+    ($( { $name:ident, $ascii:literal, $display:literal, {$($colors:tt)*} $(, $serialize:literal )? } ),* ,) => {
```

Comment:
Is it necessary for the pattern to be `:tt`? This is the least specific pattern possible, and tends to break pattern matching. Generally, I've only seen `$( $tokens:tt )*` used at the *end* of macro rule definitions, to mark optional extra tokens. In fact, I'm a bit surprised this isn't breaking pattern matching, since `}` is a token. I would assume that `{$($colors:tt)*}` would create ambiguity on whether the closing `}` is *part* of the stream of tokens or *after* the stream of tokens.
I would expect `$colors:expr` to be sufficient, since both a `vec!` declaration and a struct declaration should match an expression.

Reply:
@spenserblack thank you for your review, I greatly value any feedback on this. I simply don't know what a good answer looks like for this. What would you advise?

You are correct.  ```:tt``` is not required and other approaches exist. I wasn't sure what a good solution looks like for this and selected to keep everything inside macro expansion because that is where I think the problem begins.

AFAIK, the problem is when ```$colors``` is ```expr``` in the parent macro both ```vec![ Color::Red]``` and ```Color { ... }``` are matches for ```expr```. Unfortunately, the parent match provides a singular match for nested macros like ```define_colors```. The nested macro won't match ```vec!``` because the literal token ```vec!``` isn't available (its been flattened), instead ```vec! ...``` matches ```expr``` in ```define_colors``` and is processed like ```Colors``` which is wrong.

An alternative I found online uses ```downcast_ref``` to do type checking after macro expansion. This approach has less impact on the original table, but delays separating ```vec``` and ```Colors``` until type checking is possible. I am not sure if that is compile time or runtime checking in this case.

``` rust
macro_rules! define_colors {
    ( $color:expr ) => {
        if let Some(vector) = (&$color as &dyn Any).downcast_ref::<Vec<Color>>() {
            Colors { basic_colors: vector.clone(), true_colors: None }
        }
        else if let Some(colors) = (&$color as &dyn Any).downcast_ref::<Colors>() {
            colors.clone()
        }
        else {
            Colors { basic_colors: vec![ Color::White ], true_colors: None }
        }
    };
}
```

Reply:
I see.
An alternative is to use `define_colors!` inside the `define_languages!` macro invocation. Here's a very simple example.

```rust
macro_rules! define_colors {
    ( [ $( color:expr ),+ ] ) => { Colors { basic_colors: vec![$( $color ),+ ], true_colors: None } };
    ( $color:expr ) => { $color };
}

macro_rules! define_languages {                                               
     ($( { $name:ident, $ascii:literal, $display:literal, $colors:expr $(, $serialize:literal )? } ),* ,) => {
        /* macro definition */
    }
}

define_languages! {
    { C, "c.ascii", "C", define_colors!( [Color::Cyan, Color::Blue] ) },
}
```
I slightly changed the first rule of `define_colors!` so that it doesn't require `vec!`. Since it will always be added, might as well let the macro deal with it.

Reply:
@spenserblack , thank you and I agree moving define_colors into the invocation is an improvement.